feat(prebuilt/cloud-sql): Add clone instance tool for cloud sql

nandinisaini-google opened the pull request:
## Description

---
This pull request adds a new tool, cloud-sql-clone-instance, which enables cloning a Cloud SQL instance from the toolbox using the Cloud SQL Admin API. The tool supports both standard cloning and point-in-time recovery (PITR). It also supports specifying preferred zones for cloned instances via the preferredZone and preferredSecondaryZone fields.

Key Features:
Instance Cloning: The tool allows you to clone a Cloud SQL instance by specifying the source and destination instance names.
Point-in-Time Recovery (PITR): By providing a pointInTime timestamp, you can create a clone of an instance as it existed at a specific moment.
High Availability Configuration: The preferredZone and preferredSecondaryZone parameters allow you to configure the cloned instance for high availability. 

Tested: 
<img width="1182" height="446" alt="Screenshot 2025-11-11 at 12 21 47 PM" src="https://github.com/user-attachments/assets/7f39a5a3-3967-43d0-8041-f1d47b4fbcd9" />


## PR Checklist

> Thank you for opening a Pull Request! Before submitting your PR, there are a
> few things you can do to make sure it goes smoothly:

- [x] Make sure you reviewed
  [CONTRIBUTING.md](https://github.com/googleapis/genai-toolbox/blob/main/CONTRIBUTING.md)
- [x] Make sure to open an issue as a
  [bug/issue](https://github.com/googleapis/genai-toolbox/issues/new/choose)
  before writing your code! That way we can discuss the change, evaluate
  designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)
- [ ] Make sure to add `!` if this involve a breaking change

🛠️ Fixes #1915

juhiacharya-google commented on docs/en/resources/tools/cloudsql/cloudsqlcloneinstance.md:
```diff
@@ -0,0 +1,55 @@
+
+| **parameter**              | **type** | **required** | **description**                                                                 |
+| -------------------------- | :------: | :----------: | ------------------------------------------------------------------------------- |
+| projectId                  | string   | true         | The project ID.                                                                |
```
Should we use 'project' to be consistent with the create tool?

prernakakkar-google commented:
Fix the title of pull request

nandinisaini-google commented on docs/en/resources/tools/cloudsql/cloudsqlcloneinstance.md:
```diff
@@ -0,0 +1,55 @@
+- You can optionally specify the `zone` parameter to set the zone for the cloned instance. If omitted, the zone of the source instance will be used.
+- You can optionally specify the `preferredZone` and `preferredSecondaryZone` (only in REGIONAL instances) to set the preferred zones for the cloned instance. These are useful for high availability (HA) configurations. If omitted, defaults will be used based on the source instance.
+
+## See Also
```
Done

nandinisaini-google commented:
> Fix the title of pull request

Done

nandinisaini-google commented on internal/tools/cloudsql/cloudsqlcloneinstance/cloudsqlcloneinstance.go:
```diff
@@ -0,0 +1,188 @@
+
+	description := cfg.Description
+	if description == "" {
+		description = "Clones a Cloud SQL instance."
```
Done

prernakakkar-google commented on internal/tools/cloudsql/cloudsqlcloneinstance/cloudsqlcloneinstance.go:
```diff
@@ -0,0 +1,187 @@
+
+	project, ok := paramsMap["projectId"].(string)
+	if !ok {
+		return nil, fmt.Errorf("error casting 'project' parameter: %v", paramsMap["projectId"])
```
projectId

prernakakkar-google commented on docs/en/resources/tools/cloudsql/cloudsqlcloneinstance.md:
```diff
@@ -0,0 +1,56 @@
+
+| **parameter**              | **type** | **required** | **description**                                                                 |
+| -------------------------- | :------: | :----------: | ------------------------------------------------------------------------------- |
+| project                    | string   | true         | The project ID.                                                                 |
```
projectId

nandinisaini-google replied:
Done

prernakakkar-google commented on docs/en/resources/tools/cloudsql/cloudsqlcloneinstance.md:
```diff
@@ -0,0 +1,56 @@
+  clone-instance-tool:
+    kind: cloud-sql-clone-instance
+    source: cloud-sql-admin-source
+    description: "Clones a Cloud SQL instance."
```
should the example cover optional parameters like PITR. for clarity

prernakakkar-google commented:
/gcbrun

prernakakkar-google commented:
Please correct the title to (prebuilt/<kind-name>. Please fill in sections such as PR checklist in description.

prernakakkar-google reviewed (changes requested):
Please address the comments

prernakakkar-google commented on cmd/root_test.go:
```diff
@@ -1394,7 +1394,7 @@ func TestPrebuiltTools(t *testing.T) {
 					Name:      "cloud_sql_mssql_admin_tools",
-					ToolNames: []string{"create_instance", "get_instance", "list_instances", "create_database", "list_databases", "create_user", "wait_for_operation"},
+					ToolNames: []string{"create_instance", "get_instance", "list_instances", "create_database", "list_databases", "create_user", "wait_for_operation", "clone_instance"},
 				},
```
nit:  put this after list_instances here and elsewhere

nandinisaini-google replied:
Done

averikitsch commented on docs/en/resources/tools/cloudsql/cloudsqlcloneinstance.md:
```diff
@@ -0,0 +1,79 @@
+description: "Clone a Cloud SQL instance."
+---
+
+The `cloud-sql-clone-instance` tool clones a Cloud SQL instance using the Cloud SQL Admin API.
```
Are there any specific permissions required to use this tool?

nandinisaini-google replied:
Not anythig specific. The user or service account just needs to have admin role.

averikitsch commented on internal/tools/cloudsql/cloudsqlcloneinstance/cloudsqlcloneinstance.go:
```diff
@@ -0,0 +1,187 @@
+
+	description := cfg.Description
+	if description == "" {
+		description = "Clone an existing Cloud SQL instance into a new instance. Required fields are projectId (the GCP project ID), sourceInstanceName (the name of the instance to clone), and destinationInstanceName (the name to assign to the new instance). Optional fields include pointInTime (a timestamp for point-in-time cloning), preferredZone (the desired zone for the new instance), and preferredS ...
```
This description focuses on the inputs but the inputs are already specified by the tool. Can we iterate on this description to either be shortened to save context token or have specific details to improve usage?

nandinisaini-google commented on internal/tools/cloudsql/cloudsqlcloneinstance/cloudsqlcloneinstance.go:
```diff
@@ -0,0 +1,187 @@
+	}
+
+	allParameters := tools.Parameters{
+		tools.NewStringParameter("projectId", "The project ID of the project containing the instance to be cloned."),
```
Added it in the description

nandinisaini-google commented on internal/tools/cloudsql/cloudsqlcloneinstance/cloudsqlcloneinstance.go:
```diff
@@ -0,0 +1,187 @@
+	}
+
+	allParameters := tools.Parameters{
+		tools.NewStringParameter("projectId", "The project ID of the project containing the instance to be cloned. Use default project id of the source instance if specified."),
```
Done

nandinisaini-google commented on internal/tools/cloudsql/cloudsqlcloneinstance/cloudsqlcloneinstance.go:
```diff
@@ -0,0 +1,187 @@
+	}
+
+	allParameters := tools.Parameters{
+		tools.NewStringParameter("projectId", "The project ID of the project containing the instance to be cloned. Use default project id of the source instance if specified."),
```
Done

averikitsch commented:
/gcbrun

nandinisaini-google commented:
> Hi all, there's some updates within the code of Toolbox during our last release. If you have a PR in progress, please see the following updates: Config is now embedded in the Resource (e.g. Source / Tool / Auth Services) struct [PR reference here](https://github.com/googleapis/genai-toolbox/pull/1864/files#diff-bcf3320ab094f274841d7b075e7505ffa7d9cdac653aa6fb52c1df33c02553c1) struct now have a new ToConfig() function [PR reference here](https://github.com/googleapis/genai-toolbox/pull/1864/files#diff-bcf3320ab094f274841d7b075e7505ffa7d9cdac653aa6fb52c1df33c02553c1) parameters is now moved from internal/tools to internal/util [PR here](https://github.com/googleapis/genai-toolbox/pull/1907)
> 
> Please reach out if you have any questions :) Thank you!!

Updated the PR with the above mentioned changes, @averikitsch let me know if any further changes are required. Thankyou

nandinisaini-google commented:
> @nandinisaini-google can you take a look at the test issues.
<img width="1066" height="360" alt="Screenshot 2025-11-18 at 11 50 26 AM" src="https://github.com/user-attachments/assets/5f24c6e4-391a-4da8-9f49-5f2d1d4649a4" />

It looks like the cloudsql tests passed. The other test failure seem unrelated.

prernakakkar-google commented:
/gcbrun

prernakakkar-google commented on internal/prebuiltconfigs/tools/cloud-sql-mssql-admin.yaml:
```diff
@@ -39,6 +39,10 @@ tools:
   wait_for_operation:
     kind: cloud-sql-wait-for-operation
     source: cloud-sql-admin-source
+    multiplier: 4
```
If you are making this change please validate other tools as well

averikitsch commented:
/gcbrun

averikitsch commented:
/gcbrun

averikitsch commented:
/gcbrun
